### PR TITLE
feat: Add interactive pagination to 'mesheryctl exp model list'

### LINKED ISSUE #9625: [mesheryctl] New command: `exp components [list|view|search]`
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

### Current Behavior
Create an `exp components [list|view|search]` command similar to that of `exp model` command.

---

### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](http://discuss.meshery.io) and [Layer5 Community Slack](https://slack.meshery.io)

### PR DESCRIPTION
**Pull request by RipulHandoo**
**Notes for Reviewers**
This PR fixes #9625  
And 
1. This PR adds interactive pagination to 'mesheryctl exp model list.'
2. Adds a count flag to retrieve the total number of models.
3. Displays a message when a user searches for a model not in the list.

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [X] Yes, I signed my commits.

### CODE REVIEW

**Comment by saurabh100ni** [comment]
Hi @RipulHandoo 
Thanks for your PR! I encourage you to share this in the Meshery Development Meeting tomorrow at 8:30 PM IST. Share your screen, and walk us through it. It's cool to take and share feedback. Check the [Doc](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit) here to include this PR.

You're welcome at the meeting

Check to [community calendar](https://layer5.io/community/calendar) for more details of meetings at Layers.

File: mesheryctl/internal/cli/root/system/component.go [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+					return err
+				}
+
+				defer func() {
+					_ = keyboard.Close()
+				}()
+
+				event := <-keysEvents
+				if event.Err != nil {
+					utils.Log.Error(fmt.Errorf("unable to capture keyboard events"))
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:165]:
```suggestion
					utils.Log.Error(fmt.Errorf("unable to capture keyboard events"))
```
Please use the `meshkit` error.
If `meshkit logger` is used to log non-meshkit error it results in panic.

File: mesheryctl/internal/cli/root/system/component.go [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+			utils.Log.Error(err)
+			return err
+		}
+
+		// defers the closing of the response body after its use, ensuring that the resources are properly released.
+		defer resp.Body.Close()
+
+		data, err := io.ReadAll(resp.Body)
+		if err != nil {
+			utils.Log.Error(err)
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:255]: Same here.

File: mesheryctl/internal/cli/root/system/component.go [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+		} else {
+			selectedComponent = selectComponentPrompt(componentResponse.Components)
+		}
+
+		var output []byte
+
+		// user may pass flag in lower or upper case but we have to keep it lower
+		// in order to make it consistent while checking output format
+		outFormatFlag = strings.ToLower(outFormatFlag)
+		if outFormatFlag == "yaml" {
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:282]: Please address the case when "yml" is passed.

File: mesheryctl/internal/cli/root/system/component.go [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+	"gopkg.in/yaml.v2"
+)
+
+// represents the mesheryctl exp components list command
+var listComponentCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List registered components",
+	Long:  "List all components registered in Meshery Server",
+	Example: `
+	// View list of components
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:45]: How to list components of a particular model?

File: mesheryctl/internal/cli/root/system/component.go [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+		for _, component := range componentsResponse.Components {
+			if len(component.DisplayName) > 0 {
+				rows = append(rows, []string{component.Model.Name, component.Kind, component.APIVersion})
+			}
+		}
+
+		if len(rows) == 0 {
+			// if no component is found
+			fmt.Println("No components(s) found")
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:128]: whiteBoardPrinter.Print("No components(s) found")?

File: mesheryctl/internal/cli/root/system/component.go [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+				if event.Key == keyboard.KeyEnter || event.Key == keyboard.KeyArrowDown {
+					startIndex += maxRowsPerPage
+					endIndex = min(len(rows), startIndex+maxRowsPerPage)
+				}
+
+				if startIndex >= len(rows) {
+					break
+				}
+
+				whiteBoardPrinter.Println("Press Enter or ↓ to continue, Esc or Ctrl+C to exit")
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:182]:
```suggestion
				whiteBoardPrinter.Println("Press Enter or ↓ to continue, Esc or Ctrl+C to exit")
```
Ctrl/cmd to adress OS X users.

File: mesheryctl/internal/cli/root/system/component.go [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+			}
+		}
+		return nil
+	},
+}
+
+// represents the mesheryctl exp components view [component-name] subcommand.
+var viewComponentCmd = &cobra.Command{
+	Use:   "view",
+	Short: "view registered components",
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:192]:
```suggestion
	Short: "view registered component",
```
Given the fact this command only outputs details for single component right?

File: mesheryctl/internal/cli/root/system/component.go [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+// ComponentsCmd represents the mesheryctl exp components command
+var ComponentsCmd = &cobra.Command{
+	Use:   "components",
+	Short: "View list of components and detail of components",
+	Long:  "View list of components and detailed information of a specific component",
+	Example: `
+// To view list of components
+mesheryctl exp components list
+// To view a specific component
+mesheryctl exp components view [component-name]
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/component.go:402]: How is this different from 
https://github.com/meshery/meshery/pull/10118/files#diff-09fcaa361285a17c501b91b394c9c6ef119700f161e9ac10f7eb6572bde16336R190
?

File: mesheryctl/pkg/utils/formatter.go [hunk 9]
```diff
@@ -46,63 +41,4 @@ func SetupMeshkitLogger(debugLevel bool, output io.Writer) {
...
-type Paginator struct {
-	data                 []string
-	lastVisibleItemIndex int
-	whiteBgPrinter       *color.Color
-	userMsg              string
-}
-
-func NewPaginator(header []string, rows [][]string) *Paginator {
-	var buf bytes.Buffer
-	renderedTableString := writeTableToBuffer(&buf, header, rows)
```
Comment (MUzairS15) [inline mesheryctl/pkg/utils/formatter.go:60]: Why these have been deleted?
If deleted, where they have been moved to?

File: mesheryctl/internal/cli/root/system/model.go [hunk 10]
```diff
@@ -40,9 +41,12 @@ var (
...
+
+	// Maximum number of rows to be displayed in a page
+	maxRowsPerPage = 25
 )
 
-// represents the `mesheryctl exp model list` subcommand.
+// represents the mesheryctl exp model list subcommand.
 var listModelCmd = &cobra.Command{
 	Use:   "list",
 	Short: "list models",
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/model.go:52]:
```suggestion
	Short: "List registered models",
```

**Review by MUzairS15 — CHANGES_REQUESTED** [review]
Thank you for moving forward here.

File: mesheryctl/internal/cli/root/system/component.go (hunk 5 above)
Comment (RipulHandoo) [reply mesheryctl/internal/cli/root/system/component.go]: Done. ✅

File: mesheryctl/internal/cli/root/system/component.go (hunk 6 above)
Comment (RipulHandoo) [reply mesheryctl/internal/cli/root/system/component.go]: Done. ✅

File: mesheryctl/internal/cli/root/system/component.go (hunk 7 above)
Comment (RipulHandoo) [reply mesheryctl/internal/cli/root/system/component.go]: Done. ✅

File: mesheryctl/internal/cli/root/system/component.go (hunk 3 above)
Comment (RipulHandoo) [reply mesheryctl/internal/cli/root/system/component.go]: Done. ✅

File: mesheryctl/internal/cli/root/system/model.go (hunk 10 above)
Comment (RipulHandoo) [reply mesheryctl/internal/cli/root/system/model.go]: Done. ✅

File: mesheryctl/pkg/utils/formatter.go (hunk 9 above)
Comment (RipulHandoo) [reply mesheryctl/pkg/utils/formatter.go:60]: They were created for the pagination, it was redundant code. Also it was only showing one new model/component at a time. Now we are using the once method to print table.

File: mesheryctl/internal/cli/root/system/component.go (hunk 8 above)
Comment (RipulHandoo) [reply mesheryctl/internal/cli/root/system/component.go:403]: This is the entry point for the components command. Here we are showing the example of how to use `mesheryctl exp component` command.

**Comment by leecalcote** [comment]
Please look into the failing Mesheryctl build check.

**Comment by leecalcote** [comment]
@RipulHandoo, the mesheryctl build check is failing...


I'm rebasing the branch now.

**Comment by RipulHandoo** [comment]
> @RipulHandoo, the mesheryctl build check is failing...
> 
> I'm rebasing the branch now.

@leecalcote, the failing tests are the same as I mentioned previously [here](https://github.com/meshery/meshery/pull/9726#issuecomment-1906463735).

**Pull request merged**